connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java

```diff
@@ -1123,6 +1125,24 @@ private void processTasksCommitRecord(String connectorName, SchemaAndValue value
                     connectorTaskConfigGenerations.compute(connectorName, (ignored, generation) -> generation != null ? generation + 1 : 0);
                 }
                 inconsistent.remove(connectorName);
+
+                Map<String, String> rawConnectorConfig = connectorConfigs.get(connectorName);
+                Map<String, String> appliedConnectorConfig;
+                if (configTransformer != null) {
+                    try {
+                        appliedConnectorConfig = configTransformer.transform(rawConnectorConfig);
```

Comment:
Let's say that there's 1000 connector configs in the config topic. Today I think the configTransformer only transforms that config when it's in-use (like a connector being started, fenced, updated, etc). That should only be a handful of times after the startup completes.

With the code as-written, I think this would make 1000 configTransformer requests, including ones on out-of-date configurations. If the ConfigProvider is making a network request, this has the potential to be slow/expensive, and generate a spike of load that wasn't there before.

Do you think there's a simple way to avoid this burst of ConfigProvider calls during startup? The built-in AK providers can handle the burst safely, but I wonder if there are third-party providers that could have bad side-effects.

Reply:
Great catch! I've tweaked the PR to lazily apply config providers only when we actually make use of applied connector configs (i.e., when checking to see if we should publish a new set of task configs).

This does come with a small downside in that config transformation takes place a little bit later than it could, so it may take longer for a new set of task configs to be published if a connector config changes as a result of new values being applied by its config provider. But I'm hopeful that this can be left as a follow-up task if the basic approach in my latest commit looks good. Eager to hear your thoughts!